File: go/tests/civic_board_test.go
// civic_board_test.go
//
// Verifies the CircleAI.Civic port (civic_board.go): report/resolve with
// open-issue filtering (sorted), reps-for-district (case-insensitive, nil-safe),
// and upcoming events ordering.

package circleai_test

import (
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestCivic_IssuesAndReps(t *testing.T) {
	b := circleai.NewInMemoryCivicBoard()
	b.Report(circleai.CivicIssue{IssueId: "i2", Category: "roads", Description: "pothole", Status: "Open"})
	b.Report(circleai.CivicIssue{IssueId: "i1", Category: "water", Description: "leak", Status: "Open"})
	b.Report(circleai.CivicIssue{IssueId: "i3", Category: "power", Description: "outage", Status: "Open"})
	if err := b.Resolve("i3", "Resolved"); err != nil {
		t.Fatalf("resolve: %v", err)
	}
	if err := b.Resolve("ghost", "Resolved"); err == nil {
		t.Fatalf("resolve unknown must error")
	}
	open := b.OpenIssues()
	if len(open) != 2 || open[0].IssueId != "i1" || open[1].IssueId != "i2" {
		t.Fatalf("open issues sorted failed: %+v", open)
	}

	d := "Ward 5"
	b.AddRep(circleai.Representative{RepId: "r1", Name: "A", Office: "Councillor", District: &d})
	b.AddRep(circleai.Representative{RepId: "r2", Name: "B", Office: "MP"}) // nil district
	reps := b.RepsForDistrict("ward 5")
	if len(reps) != 1 || reps[0].RepId != "r1" {
		t.Fatalf("reps-for-district failed: %+v", reps)
	}
}

func TestCivic_UpcomingEvents(t *testing.T) {
	b := circleai.NewInMemoryCivicBoard()
	now := time.Now().UTC()
	b.Schedule(circleai.CivicEvent{EventId: "e1", Title: "Later", AtUtc: now.Add(48 * time.Hour), Location: "Hall", Audience: "public"})
	b.Schedule(circleai.CivicEvent{EventId: "e2", Title: "Soon", AtUtc: now.Add(24 * time.Hour), Location: "Hall", Audience: "public"})
	b.Schedule(circleai.CivicEvent{EventId: "e3", Title: "Past", AtUtc: now.Add(-24 * time.Hour), Location: "Hall", Audience: "public"})

	up := b.UpcomingEvents()
	if len(up) != 2 || up[0].EventId != "e2" || up[1].EventId != "e1" {
		t.Fatalf("upcoming events ordered failed: %+v", up)
	}
}

File: typescript/tests/civic_board.test.ts
// civic_board.test.ts
// Verifies the CircleAI.Civic port: report + resolve, open issues, reps by
// district, upcoming events.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  InMemoryCivicBoard,
  CivicDomainContext,
  civicIssue,
  representative,
  civicEvent,
} from "../src/civic/index";

describe("InMemoryCivicBoard", () => {
  it("reports issues and lists those not resolved", () => {
    const b = new InMemoryCivicBoard();
    b.report(civicIssue("i1", "Water", "Burst pipe", -26, 28, new Date("2026-01-01T00:00:00Z"), "Open"));
    b.report(civicIssue("i2", "Roads", "Pothole", -26, 28, new Date("2026-01-01T00:00:00Z"), "Resolved"));
    b.report(civicIssue("i3", "Power", "Outage", -26, 28, new Date("2026-01-01T00:00:00Z"), "InProgress"));
    assert.deepEqual(
      b.openIssues().map((i) => i.issueId).sort(),
      ["i1", "i3"],
    );
  });

  it("resolve updates status (and hides from open) or throws", () => {
    const b = new InMemoryCivicBoard();
    b.report(civicIssue("i1", "Water", "Burst pipe", -26, 28, new Date("2026-01-01T00:00:00Z"), "Open"));
    b.resolve("i1", "Resolved");
    assert.equal(b.openIssues().length, 0);
    assert.throws(() => b.resolve("ghost", "Resolved"), /Unknown issue ghost/);
  });

  it("finds reps for a district case-insensitively", () => {
    const b = new InMemoryCivicBoard();
    b.addRep(representative("r1", "Ann", "Councillor", "ann@x", "Ward 5"));
    b.addRep(representative("r2", "Ben", "MP", "ben@x", null));
    b.addRep(representative("r3", "Cy", "Councillor", "cy@x", "ward 5"));
    assert.deepEqual(
      b.repsForDistrict("WARD 5").map((r) => r.repId).sort(),
      ["r1", "r3"],
    );
  });

  it("lists upcoming events ascending", () => {
    const b = new InMemoryCivicBoard();
    const future1 = new Date(Date.now() + 86_400_000);
    const future2 = new Date(Date.now() + 172_800_000);
    b.schedule(civicEvent("e1", "Meeting", future2, "Hall", "Public"));
    b.schedule(civicEvent("e2", "Cleanup", future1, "Park", "Public"));
    b.schedule(civicEvent("e3", "Old", new Date(Date.now() - 86_400_000), "Hall", "Public"));
    assert.deepEqual(
      b.upcomingEvents().map((e) => e.eventId),
      ["e2", "e1"],
    );
  });

  it("domain context exposes prompt + compliance + tools", () => {
    assert.ok(CivicDomainContext.systemPromptSnippet.includes("[DOMAIN: Civic]"));
    assert.deepEqual(CivicDomainContext.complianceFlags, [
      "PAJA",
      "PAIA",
      "Constitution_RSA",
      "Municipal_Systems_Act",
      "POPIA",
    ]);
    assert.deepEqual(CivicDomainContext.suggestedTools, ["government_portals", "document_editor", "map", "web_search"]);
  });
});
